spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/reactor/TraceReactorAutoConfiguration.java

```diff
@@ -194,27 +213,122 @@ public void postProcessBeanFactory(ConfigurableListableBeanFactory beanFactory)
+		Boolean decorateHooks = environment
+				.getProperty("spring.sleuth.reactor.decorate-hooks", Boolean.class);
+		boolean hasQueueWrapper = ReflectionUtils.findMethod(Hooks.class,
+				"addQueueWrapper", String.class, Function.class) != null;
+		if (!hasQueueWrapper && decorateHooks != null) {
+			log.warn(
+					"You have explicitly turned on the decorate hooks option but you're using an old version of Reactor. Please upgrade to the latest Boot version (at least 2.3.9.RELEASE). Will fall back to the previous reactor instrumentation mode");
 		}
 		else {
-			if (log.isTraceEnabled()) {
-				log.trace("Decorating onLast operator instrumentation");
+			decorateHooks = decorateHooks != null ? decorateHooks : Boolean.TRUE;
+		}
+		if (Boolean.TRUE.equals(decorateHooks) && hasQueueWrapper) {
+			addQueueWrapper(springContext);
+		}
+		else {
+			boolean decorateOnEach = environment.getProperty(
+					"spring.sleuth.reactor.decorate-on-each", Boolean.class, true);
+			if (decorateOnEach) {
+				if (log.isTraceEnabled()) {
+					log.trace("Decorating onEach operator instrumentation");
+				}
+				Hooks.onEachOperator(SLEUTH_TRACE_REACTOR_KEY,
+						scopePassingSpanOperator(springContext));
 			}
-			Hooks.onLastOperator(SLEUTH_TRACE_REACTOR_KEY,
-					scopePassingSpanOperator(springContext));
+			else {
+				if (log.isTraceEnabled()) {
+					log.trace("Decorating onLast operator instrumentation");
+				}
+				Hooks.onLastOperator(SLEUTH_TRACE_REACTOR_KEY,
+						scopePassingSpanOperator(springContext));
+			}
+		}
+		decorateScheduler(springContext);
+	}
+
+	static void addQueueWrapper(ConfigurableApplicationContext springContext) {
+		Hooks.addQueueWrapper(SLEUTH_TRACE_REACTOR_KEY,
+				queue -> traceQueue(springContext, queue));
+	}
+
+	@Override
+	public void close() throws IOException {
+		if (log.isTraceEnabled()) {
+			log.trace("Cleaning up hooks");
 		}
-		Schedulers.setExecutorServiceDecorator(
+		Hooks.resetOnEachOperator(SLEUTH_TRACE_REACTOR_KEY);
+		Hooks.resetOnLastOperator(SLEUTH_TRACE_REACTOR_KEY);
+		Schedulers.resetOnScheduleHook(
+				TraceReactorAutoConfiguration.SLEUTH_REACTOR_EXECUTOR_SERVICE_KEY);
+	}
+
+	private static void decorateScheduler(ConfigurableApplicationContext springContext) {
+		Schedulers.onScheduleHook(
 				TraceReactorAutoConfiguration.SLEUTH_REACTOR_EXECUTOR_SERVICE_KEY,
-				(scheduler,
-						scheduledExecutorService) -> new TraceableScheduledExecutorService(
-								springContext, scheduledExecutorService));
+				ReactorSleuth.scopePassingOnScheduleHook(springContext));
+	}
+
+	private static Queue<?> traceQueue(ConfigurableApplicationContext springContext,
+			Queue<?> queue) {
+		if (!springContext.isActive()) {
+			return queue;
+		}
+		CurrentTraceContext currentTraceContext = springContext
+				.getBean(CurrentTraceContext.class);
+		@SuppressWarnings("unchecked")
+		Queue<Envelope> envelopeQueue = (Queue<Envelope>) queue;
+		return new AbstractQueue<Object>() {
```

Comment:
Would it make sense to move this into a separate, package-private class (inject the `currentTraceContext`)?
That way this would be testable in itself and maybe a little more readable.